lib/net/packets.js

```diff
@@ -957,6 +970,663 @@ class NotFoundPacket extends InvPacket {
+    this.type = exports.types.GETCFHEADERS;
+
+    this.filterType = filterType;
+    this.startHeight = startHeight || 0;
+    this.stopHash = stopHash || consesus.ZERO_HASH;
+  }
+
+  /**
+   * Get serialization size.
+   * @returns {Number}
+   */
+
+  getSize() {
+    return 37;
+  }
+
+  /**
+   * Serialize getcfheaders packet to writer.
+   * @param {BufferWriter} bw
+   */
+
+  toWriter(bw) {
+    bw.writeU8(this.filterType);
+    bw.writeU32(this.startHeight);
+    bw.writeHash(this.stopHash);
+
+    return bw;
+  }
+
+  /**
+   * Serialize getcfheaders packet.
+   * @returns {Buffer}
+   */
+
+  toRaw() {
+    const size = this.getSize();
+    return this.toWriter(bio.write(size)).render();
+  }
+
+  /**
+   * Inject properties from buffer reader.
+   * @private
+   * @param {BufferReader} br
+   * @returns {GetCFHeadersPacket}
+   */
+
+  fromReader(br) {
+    this.filterType = br.readU8();
+    this.startHeight = br.readU32();
+    this.stopHash = br.readHash();
+
+    return this;
+  }
+
+  /**
+   * Inject properties from serialized data.
+   * @private
+   * @param {Buffer} data
+   * @returns {GetCFHeadersPacket}
+   */
+
+  fromRaw(data) {
+    return this.fromReader(bio.read(data));
+  }
+
+  /**
+   * Instantiate getcfheaders packet from buffer reader.
+   * @param {BufferReader} br
+   * @returns {GetCFHeadersPacket}
+   */
+
+  static fromReader(br) {
+    return new this().fromReader(br);
+  }
+
+  /**
+   * Instantiate getcfheaders packet from serialized data.
+   * @param {Buffer} data
+   * @returns {GetCFHeadersPacket}
+   */
+
+  static fromRaw(data) {
+    return new this().fromRaw(data);
+  }
+}
+
+/**
+ * CFHeaders Packet
+ * @extends Packet
+ * @property {Number} filterType
+ * @property {Hash?} stopHash
+ * @property {Hash} previousFilterHeader
+ * @property {(Hash[])?} filterHashes
+ */
+
+class CFHeadersPacket extends Packet {
+  /**
+   * Create a `cfheaders` packet.
+   * @constructor
+   * @param {Number} filterType
+   * @param {Hash?} stopHash
+   * @param {Hash} previousFilterHeader
+   * @param {(Hash[])?} filterHashes
+   */
+
+  constructor(filterType, stopHash, previousFilterHeader, filterHashes) {
+    super();
+
+    this.cmd = 'cfheaders';
+    this.type = exports.types.CFHEADERS;
+
+    this.filterType = filterType;
+    this.stopHash = stopHash || consesus.ZERO_HASH;
+    this.previousFilterHeader = previousFilterHeader || consensus.ZERO_HASH;
+    this.filterHashes = filterHashes || [];
+  }
+
+  /**
+   * Get serialization size.
+   * @returns {Number}
+   */
+
+  getSize() {
+    let size = 65;
+    size += encoding.sizeVarint(this.filterHashes.length);
+    size += this.filterHashes.length * 32;
+    return size;
+  }
+
+  /**
+   * Serialize cfheaders packet to writer.
+   * @param {BufferWriter} bw
+   */
+
+  toWriter(bw) {
+    bw.writeU8(this.filterType);
+    bw.writeHash(this.stopHash);
+    bw.writeHash(this.previousFilterHeader);
+    bw.writeVarint(this.filterHashes.length);
```

Comment:
like `fromReader()` I think we should check this length value before proceeding.

Reply:
Also I think `CFCheckptPacket()` needs size checks in both read and write functions...?